### PR TITLE
Making facts understand which spaces depend on them, and be able to be defined by spaces.

### PR DESCRIPTION
**Pull request by mmcdermott**

### CODE REVIEW

File: app/models/space.rb [hunk 1]
```diff
@@ -17,12 +18,14 @@ class Space < ActiveRecord::Base
...
 
   after_initialize :init
-  after_save :identify_user
+  after_save :identify_user, :update_facts_used!
   after_destroy :identify_user
 
   scope :is_private, -> { where(is_private: true) }
   scope :is_public, -> { where(is_private: false) }
   scope :uncategorized_since, -> (date) { where 'categorized IS NOT true AND DATE(created_at) >= ?', date }
+  scope :defines_fact, -> { where('defined_facts_count > 0') }
```
Comment (OAGr) [inline app/models/space.rb]: maybe ``defines_facts``? (plural)

File: app/models/fact.rb [hunk 2]
```diff
@@ -25,8 +31,16 @@ def take_checkpoint(author)
     return checkpoint
   end
 
+  def dependent_fact_defining_spaces
```
Comment (OAGr) [inline app/models/fact.rb]: maybe space_inputs and then elsewhere space_outputs?

File: app/models/space.rb [hunk 3]
```diff
@@ -9,6 +9,7 @@ class Space < ActiveRecord::Base
   has_many :copies, class_name: 'Space', foreign_key: 'copied_from_id'
   has_many :checkpoints, class_name: 'SpaceCheckpoint', dependent: :destroy
   has_many :calculators, dependent: :destroy
+  has_many :defined_facts, foreign_key: 'defining_space_id', class_name: 'Fact', dependent: :destroy
```
Comment (OAGr) [inline app/models/space.rb]: what about calling this fact_outputs?

**Comment by OAGr** [comment]
I think this makes sense, but I would like it more if we could rename used_facts ``exported_facts`` and ``imported_facts``, or maybe ``fact_ouputs`` and ``fact_inputs`` (I think imported/exported is cleaner)

File: app/models/fact.rb (hunk 2 above)
Comment (mmcdermott) [inline app/models/fact.rb]: `dependent_fact_defining_spaces` returns a list of spaces that depend on this fact that also define facts themselves, so I think `space_inputs` wouldn't make that much sense. Unless perhaps I misunderstood your suggested renaming here?

**Comment by mmcdermott** [comment]
Closed until tests passing.

**Comment by mmcdermott** [comment]
`

**Comment by mmcdermott** [comment]
Realized a necessary piece is missing.

**Comment by mmcdermott** [comment]
Closing for now as working with the client on this has made me realize a potentially simpler first pass solution.

File: app/models/space.rb [hunk 4]
```diff
@@ -59,6 +62,23 @@ def someone_else_editing?(current_user, previous_updated_at_str)
...
+
+  def get_imported_fact_ids()
+    guesstimate_expressions.map {|e| e.scan(/\$\{fact:(\d+)/) unless e.blank? }.flatten.uniq.keep_if { |e| e.present? }
+  end
+
+  def update_exported_facts_count!()
+    update_columns(exported_facts_count: exported_facts.count)
+  end
+
+  def update_imported_facts!()
```
Comment (OAGr) [inline app/models/space.rb]: remove the parens for this and above

File: app/models/space.rb [hunk 5]
```diff
@@ -59,6 +62,23 @@ def someone_else_editing?(current_user, previous_updated_at_str)
     checkpoints.last.author_id != current_user.id
   end
 
+  def guesstimate_expressions
+    return [] unless graph.present? && graph["guesstimates"].present?
+    graph["guesstimates"].map { |g| g["expression"] }
```
Comment (OAGr) [inline app/models/space.rb]: single quotes

File: app/models/space.rb [hunk 6]
```diff
@@ -234,6 +254,14 @@ def to_text
     "
   end
 
+  def increment_exported_facts_count!
+    update_columns(exported_facts_count: exported_facts_count + 1)
+  end
+
+  def decrement_exported_facts_count!
+    update_columns(exported_facts_count: exported_facts_count - 1)
```
Comment (OAGr) [inline app/models/space.rb]: let's not do this now, but we really shouldn't have to keep on using update_columns.

File: app/models/fact.rb [hunk 7]
```diff
@@ -25,8 +34,28 @@ def take_checkpoint(author)
     return checkpoint
   end
 
+  def dependent_fact_exporting_space_ids
```
Comment (OAGr) [inline app/models/fact.rb]: Remove these for now, as we discussed.

File: app/representers/organization_representer.rb [hunk 8]
```diff
@@ -30,6 +30,10 @@ class OrganizationRepresenter < Roar::Decorator
     decorator: FactRepresenter,
     if: ->(user_options:, **) { user_options[:current_user_is_member] }
 
+  collection :fact_exporting_spaces,
+    class: Space,
+    decorator: SpaceWithoutOrganizationRepresenter
```
Comment (OAGr) [inline app/representers/organization_representer.rb]: possible remove this as well

File: spec/models/fact_spec.rb [hunk 9]
```diff
@@ -82,6 +86,49 @@
...
+        before do
+          exporting_space
+          expect(exporting_space.exported_facts_count).to be 0
+        end
+
+        it { is_expected.to be_valid }
+
+        it 'should increment the exporting_space exported_facts_count upon creation' do
+          fact.save
+          exporting_space
```
Comment (OAGr) [inline spec/models/fact_spec.rb]: I believe the ``exporting_space`` line in the ``before do`` would happen here.  It seems strange if it needs to happen twice.

File: spec/models/fact_spec.rb [hunk 10]
```diff
@@ -82,6 +86,49 @@
       let (:stats) { {"mean" => 0.5, "length" => 3, "stdev" => 1} }
       it { is_expected.to_not be_valid }
     end
+
+    context 'with exported space' do
+      let (:exporting_space) { FactoryGirl.create(:space) }
+      let (:metric_id) { nil }
+
+      it { is_expected.to_not be_valid }
```
Comment (OAGr) [inline spec/models/fact_spec.rb]: Add a context for 'without a metric id' around that it block.

File: spec/models/fact_spec.rb [hunk 11]
```diff
@@ -82,6 +86,49 @@
...
+  end
+
+  describe '#destroy' do
+    let (:exporting_space) { FactoryGirl.create(:space) }
+    let (:metric_id) { '3' }
+    subject (:fact) { FactoryGirl.create(:fact, exporting_space: exporting_space, metric_id: metric_id) }
+
+    before do
+      exporting_space # To initialize the variable.
+      fact # To initialize the variable.
```
Comment (OAGr) [inline spec/models/fact_spec.rb]: remove comments here

File: spec/models/fact_spec.rb [hunk 12]
```diff
@@ -82,6 +86,49 @@
...
+        let (:metric_id) { '3' }
+
+        before do
+          exporting_space
+          expect(exporting_space.exported_facts_count).to be 0
+        end
+
+        it { is_expected.to be_valid }
+
+        it 'should increment the exporting_space exported_facts_count upon creation' do
```
Comment (OAGr) [inline spec/models/fact_spec.rb]: don't use the word ``should`` in these descriptions.  
``it increments the...``

File: spec/models/fact_spec.rb [hunk 13]
```diff
@@ -135,4 +182,28 @@
       end
     end
   end
+
+  describe 'dependent_fact_exporting_spaces' do
```
Comment (OAGr) [inline spec/models/fact_spec.rb]: Is this a method? if so, the name should be #dependent_fact_exporting_spaces

File: spec/models/fact_spec.rb [hunk 14]
```diff
@@ -82,6 +86,49 @@
...
+    before do
+      exporting_space # To initialize the variable.
+      fact # To initialize the variable.
+
+      expect(exporting_space.exported_facts_count).to be 1
+    end
+
+    it 'should decrement exported_facts_count' do
+      fact.destroy
+
```
Comment (OAGr) [inline spec/models/fact_spec.rb]: what's with the newlines?

File: spec/models/space_spec.rb [hunk 15]
```diff
@@ -152,4 +152,26 @@
       include_examples('copies properly')
     end
   end
+
+  describe '#get_imported_fact_ids' do
+    let(:guesstimates) { [] }
+    let(:space) { FactoryGirl.build(:space, graph: guesstimates.empty? ? nil : {'guesstimates' => guesstimates}) }
+    subject(:fact_ids) { space.get_imported_fact_ids }
+
+    it 'should return an empty array with no graph' do
```
Comment (OAGr) [inline spec/models/space_spec.rb]: add context, ``without facts`` vs, ``with facts``

File: spec/models/space_spec.rb [hunk 16]
```diff
@@ -152,4 +152,26 @@
...
+      expect(fact_ids).to be_empty
+    end
+
+    context 'graph with facts' do
+      let(:guesstimates) {[
+        {'expression'=>'=${fact:1} + ${fact:2} + ${fact:1}'},
+        {'expression'=>'=${fact:2} + ${fact:3} + ${fact:40}'},
+        {'expression'=>'=${fact:1} + ${fact:2} + fact:77'}
+      ]}
+      it 'should contain the referenced fact' do
```
Comment (OAGr) [inline spec/models/space_spec.rb]: facts*

File: app/models/fact.rb [hunk 17]
```diff
@@ -1,12 +1,21 @@
 class Fact < ActiveRecord::Base
   belongs_to :organization
+  belongs_to :exporting_space, class_name: 'Space'
```
Comment (OAGr) [inline app/models/fact.rb]: rename ``exported_from_space``

File: app/models/space.rb (hunk 4 above)
Comment (mmcdermott) [inline app/models/space.rb]: Call this `update_imported_fact_ids` and change the parameter name to `imported_fact_ids`

**Pull request merged**